packages/react-scripts/scripts/utils/verifyTypeScriptSetup.js

```diff
@@ -99,6 +116,21 @@ function verifyTypeScriptSetup() {
     }
 
     tsconfig = config;
+
+    // Get TS to parse and resolve any "extends"
+    // Calling this function also mutates the tsconfig above,
+    // adding in "include" and "exclude", but the compilerOptions remain untouched
+    const result = ts.parseJsonConfigFileContent(
+      config,
```

Comment:
This should probably be `tsconfig` for clarity. Kinda sucks that mutation happens, but I'm OK with it.

Reply:
Actually, we can't mutate this here. If we write out the `tsconfig` the file will include the `include` and `exclude` options which would be unexpected. Can we deep clone `tsconfig` before passing it to this function?

Reply:
Ah yes good point about it getting written out! Definitely need to fix that.

Yep we can deep clone it. We will likely need to create a couple extra variables then to hold on to the `include` and `exclude` that is mutated on the clone, as below we're relying on `tsconfig.include` and `tsconfig.exclude` to check.

Sound good?

Reply:
I'm not sure I follow. We should be fine to clone the entire thing without special behavior.

All checks are going to be made against this new `result` object AFAIK. Only changes (no reads) will be made against the original `tsconfig`.

Reply:
Hmm yeah let me think through this a bit...

So currently `tsconfig` is what was read from the `tsconfig.json`. When we call `ts.parseJsonConfigFileContent()` it mutates the provided `tsconfig`, and adds in the `include` and `exclude` properties that came in from extended tsconfig config files (I find this to be a strange behavior. Their API for reading the config files isn't the most intuitive). `tsconfig.compilerOptions` is still the original raw properties from the root `tsconfig.json`.

`parsedOptions` contains the `compilerOptions` that was parsed from reading in all the extended `tsconfig.json` files.

If we wanted to clone `tsconfig` before passing it into `ts.parseJsonConfigFileContent()`, it means we wouldn't know the `include` and `exclude` values below when doing the `if (tsconfig.include == null) { ... }` check, as the `include` and `exclude` could be part of one of the extended files as you can see in my screenshot in the top post.

I hope that makes sense. What do you think would be the best approach here?